api/src/main/java/io/opentelemetry/common/CleanReadableAttributes.java

```diff
@@ -0,0 +1,58 @@
+/*
+ * Copyright 2020, OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.opentelemetry.common;
+
+import java.util.List;
+import javax.annotation.Nullable;
+
+/**
+ * A read-only container for String-keyed attributes.
+ *
+ * <p>See {@link Attributes} for the public API implementation.
+ */
+public interface CleanReadableAttributes {
+
+  int size();
+
+  boolean isEmpty();
+
+  /** Iterates over all the key-value pairs of attributes contained by this instance. */
+  void forEach(AttributeConsumer consumer);
+
+  Boolean getBooleanValue(Object value);
```

Comment:
Mentioned it below, but these seem to not provide much value. But `boolean getBoolean(String key)`, now that sounds nice :)

Reply:
I think we should demand to the users the type retrieval. This way, we do not have to define what should happen if we call `getBoolean(key)` on an attribute that was stored as `long`, for example.

Reply:
I think such methods are exactly demanding users to do the type retrieval - they are explicitly picking the method to call. But it gives us the ability to provide a UX possibly better for users. If we throw an exception, it's no different than any other option. But we could also treat it as an empty value if it makes sense - I'd lean towards this given we generally don't want to crash user's apps. With it in the javadoc it's pretty straightforward.

Reply:
Yes, this is probably the use-case that seems most beneficial with having these extra helper methods. We can do some intelligent default behavior when the wrong one is called, and make that uniform and documented. And, if people don't want to use them, they can do the cast themselves and handle the bad conditions however they desire.

This is one of the main values of having the typed-keys..having a type-safe API for retrieval, because then you can't ask for a boolean value from a String-typed key.

Reply:
If we don't use primitive types and we return `null` value in case of the wrong type, I agree. It would make a better UX for users.

Reply:
One option I did consider was to have the API not give the consumer the Object value, but require them to make a `get` call based on the type and the value, which could enforce that the right type is passed in for the right method. This would necessitate, probably, doing some optimization of the `get` code path, which currently just traverses the entire list to find the key. I wasn't sure how exactly to do that. One option would be to provide an Iterator that kept the index to the currently location in the list, rather than doing it via the `forEach` and an consumer.

Reply:
I agree with Anuraag that `boolean getBoolean(String key)` etc would make a nice API for single key retrieval.

Reply:
easy to add, but wouldn't help with the iteration use-case, which I think will be the most common one, unfortunately.